### PR TITLE
Refactor optimization utilities

### PR DESCRIPTION
**Pull request by tnagler**

### CODE REVIEW

File: include/vinecopulib/misc/implementation/tools_optimization.ipp [hunk 1]
```diff
@@ -46,6 +37,91 @@ inline void Optimizer::set_controls(double initial_trust_region,
...
+    const Eigen::VectorXd &upper_bounds,
+    std::function<double(const Eigen::VectorXd&)> objective)
+{
+    check_parameters_size(initial_parameters, lower_bounds, upper_bounds);
+    size_t n_parameters = initial_parameters.size();
+    auto optimal_parameters = initial_parameters;
+    if (n_parameters > 1) {
+        //const int number_interpolation_conditions = (n_parameters + 1) *
+        //        (n_parameters + 2)/2;
+        int number_interpolation_conditions = n_parameters + 3;
```
Comment (tvatter) [inline include/vinecopulib/misc/implementation/tools_optimization.ipp]: ?

File: include/vinecopulib/misc/implementation/tools_optimization.ipp [hunk 2]
```diff
@@ -46,6 +37,91 @@ inline void Optimizer::set_controls(double initial_trust_region,
...
+        std::function<double(long, const double*)> f =
+            [objective, this](long n, const double *x) {
+            Eigen::Map<const Eigen::VectorXd> par(x, n);
+            this->objective_calls_++;
```
Comment (tvatter) [inline include/vinecopulib/misc/implementation/tools_optimization.ipp]: What does this do?

File: include/vinecopulib/bicop/implementation/parametric.ipp [hunk 3]
```diff
@@ -56,126 +56,126 @@ inline void ParBicop::fit(const Eigen::Matrix<double, Eigen::Dynamic, 2> &data,
...
+          newpars(0) = get_parameters()(0);
+      }
+
+      set_parameters(newpars);
+      set_loglik(optimizer.get_objective_max());
+}
+
+//! ensures that starting values are sufficiently separated from bounds
+//! @param tau kendall's tau
+inline double ParBicop::winsorize_tau(double tau) const
```
Comment (tvatter) [inline include/vinecopulib/bicop/implementation/parametric.ipp]: Funny name, winsorize would be more when tau is close to 0 or 1 no?

File: include/vinecopulib/bicop/implementation/parametric.ipp [hunk 4]
```diff
@@ -56,126 +56,126 @@ inline void ParBicop::fit(const Eigen::Matrix<double, Eigen::Dynamic, 2> &data,
...
+              return this->loglik(data, weights);
+          };
+      }
+
+      tools_optimization::Optimizer optimizer;
+      auto newpars = optimizer.optimize(initial_parameters, lb, ub, objective);
+
+      // check if fit is reasonable, otherwise increase search interval
+      // and refit
+      if (optimizer.get_objective_max() < -0.1) {
```
Comment (tvatter) [inline include/vinecopulib/bicop/implementation/parametric.ipp]: Aren't we doing the same optimization a second time for nothing for multiparameter families? If I understood the adjustment correctly, it's just for one-parameter families (for the Brent algorithm) no?

**Comment by tvatter** [comment]
Long overdue, I hated doing this initial part of the code that I didn't even want to look at it anymore..

File: include/vinecopulib/misc/implementation/tools_optimization.ipp (hunk 1 above)
Comment (tnagler) [reply include/vinecopulib/misc/implementation/tools_optimization.ipp]: I did not change this part!

File: include/vinecopulib/bicop/implementation/parametric.ipp (hunk 3 above)
Comment (tnagler) [reply include/vinecopulib/bicop/implementation/parametric.ipp]: Well it does enforce that |tau| is in [0.01, 0.9], but I'm open to other suggestions.

File: include/vinecopulib/bicop/implementation/parametric.ipp (hunk 4 above)
Comment (tnagler) [reply include/vinecopulib/bicop/implementation/parametric.ipp]: Right, I'll fix that.

File: include/vinecopulib/misc/implementation/tools_optimization.ipp (hunk 1 above)
Comment (tvatter) [reply include/vinecopulib/misc/implementation/tools_optimization.ipp]: Haha, so I have no idea why I did that ^^

File: include/vinecopulib/bicop/implementation/parametric.ipp (hunk 3 above)
Comment (tvatter) [reply include/vinecopulib/bicop/implementation/parametric.ipp]: I meant -1 and 1, my bad, and I do totally understand the logic, I just wondered about whether a better name could be found (winsorization usually doesn't do anything in the middle of the interval). In any case, I'm fine with this.

**Pull request merged**